generator/src/main/java/org/apache/kafka/message/ApiMessageTypeGenerator.java

```diff
@@ -355,28 +362,9 @@ private void generateHeaderVersion(String type) {
                 buffer.decrementIndent();
                 continue;
             }
-            if (type.equals("request") && apiKey == 7) {
-                buffer.printf("// Version 0 of ControlledShutdownRequest has a non-standard request header%n");
-                buffer.printf("// which does not include clientId.  Version 1 of ControlledShutdownRequest%n");
-                buffer.printf("// and later use the standard request header.%n");
-                buffer.printf("if (_version == 0) {%n");
-                buffer.incrementIndent();
-                buffer.printf("return (short) 0;%n");
-                buffer.decrementIndent();
-                buffer.printf("}%n");
-            }
-            ApiData data = entry.getValue();
-            MessageSpec spec;
-            if (type.equals("request")) {
-                spec = data.requestSpec;
-            } else if (type.equals("response")) {
-                spec = data.responseSpec;
-            } else {
-                throw new RuntimeException("Invalid type " + type + " for generateHeaderVersion");
-            }
-            if (spec == null) {
-                throw new RuntimeException("failed to find " + type + " for API key " + apiKey);
-            }
+            MessageSpec spec = messageSpec(type, apiKey, entry.getValue());
+            if (!spec.hasValidVersion())
+                continue;
```

Comment:
It seems we need to add `buffer.decrementIndent()` to avoid generating following code.
```java
                            case 4: // LeaderAndIsr
                                case 5: // StopReplica
                                    case 6: // UpdateMetadata
                                        case 7: // ControlledShutdown
                                            case 8: // OffsetCommit
```

Reply:
Or maybe we should skip those api kips to throw `UnsupportedVersionException` directly?

Reply:
Yeah, the intent was to skip them - this is a bug. I'll fix it and add a test.

Reply:
Done.